File: src/main/java/com/gitlab/daring/image/command/Command.java
package com.gitlab.daring.image.command;

import com.gitlab.daring.image.command.parameter.CommandParam;

import java.util.List;

import static java.util.Collections.emptyList;

@FunctionalInterface
public interface Command extends AutoCloseable {

    void execute(CommandEnv env);

    default List<CommandParam<?>> getParams() {
        return emptyList();
    }

    default boolean isEnabled(CommandEnv env) {
        return env.task.equals(env.curTask);
    }

    default boolean isCacheable() {
        return true;
    }

    default void close() {}

    @FunctionalInterface
    interface Factory {
        Command create(String[] params);
    }

}

File: src/main/kotlin/com/gitlab/daring/image/command/Command.kt
package com.gitlab.daring.image.command

import com.gitlab.daring.image.command.parameter.CommandParam

interface Command : AutoCloseable {

    fun execute(env: CommandEnv)

    val params: List<CommandParam<*>>

    val cacheable get() = true

    fun isEnabled(env: CommandEnv) = env.task == env.curTask

    override fun close() {}

}
